### PR TITLE
Add field removed_by to Agent.

### PR DESCRIPTION
**Pull request by rafbgarcia**
That is to inform who signed the agent out of the queue.

### CODE REVIEW

**Comment by bklang** [comment]
I'm not sure how this fits into the ElectricSlide model. What kind of data is `removed_by` expected to hold? Another agent's ID? I suspect there's a different/more appropriate place to catch this data in application code, rather than in this module.

**Comment by rafbgarcia** [comment]
That field will hold a String with 2 possible values: 
- The agent id
- `auto` when Talkbox automatically removes the agent from the queue

You mean that this information should be in Talkbox rather than Electric Slide?

File: lib/electric_slide/agent.rb [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 class ElectricSlide
   class Agent
-    attr_accessor :id, :address, :presence, :call, :queue
+    attr_accessor :id, :address, :presence, :call, :queue, :removed_by
```
Comment (benlangfeld) [inline lib/electric_slide/agent.rb]: I'd rather not store this on the agent, but rather pass it via a presence update only.

Comment (rafbgarcia) [inline lib/electric_slide/agent.rb]: You mean doing something like:

`agent.presence = :unavailable, removed_by: 'agent_id'`

File: spec/electric_slide/agent_spec.rb [hunk 2]
```diff
@@ -45,5 +45,21 @@ def foo
     ElectricSlide::Agent.on_presence_change { |queue, agent_call, presence| called = true }
     agent = ElectricSlide::Agent.new presence: :unavailable
     agent.presence = :busy
+
+    expect(called).to be_truthy
+  end
+
+  it 'sends `removed_by` and `old_presence` to the presence change callback' do
+    _removed_by = _old_presence = ''
```
Comment (benlangfeld) [inline spec/electric_slide/agent_spec.rb]: Leading underscores like this typically denote unused variables.

**Comment by bklang** [comment]
@rafbgarcia So if that’s the case, that if at some point ElectricSlide has to touch app-specific data, perhaps we need to do that in a generic way. For example: what happens in the future when XYZ corp decides that the need to also record the department ID of the agent being logged out? Or that some other ElectricSlide consumer wants to record the agent's spoken language?

**Comment by rafbgarcia** [comment]
@bklang what do you think of:

```ruby
    def update_presence(new_presence, extra_params = {})
      old_presence = @presence
      @presence = new_presence
      callback :presence_change, queue, @call, new_presence, old_presence, extra_params
    end
```

This way, ES will just pass the extra params back through the `presence_change` callback and won't be aware about anything application specific.

cc/@benlangfeld

File: lib/electric_slide/agent.rb [hunk 3]
```diff
@@ -14,10 +14,10 @@ def initialize(opts = {})
       @presence = opts[:presence] || :available
     end
 
-    def presence=(new_presence)
```
Comment (neildecapia) [inline lib/electric_slide/agent.rb]: Should probably remove [access to `#presence=`](https://github.com/adhearsion/electric_slide/pull/45/files#diff-0a708c3adfc0f0b8531a6430e311ce3aR5) completely. Also needs a change log entry as this is a significant API change.

**Comment by benlangfeld** [comment]
What he said ^

File: lib/electric_slide/call_queue.rb [hunk 4]
```diff
@@ -216,9 +216,10 @@ def return_agent!(*args)
 
     # Removes an agent from the queue entirely
     # @param [Agent] agent The {Agent} to be removed from the queue
+    # @param [extra_params] hash Application specific extra params
```
Comment (bklang) [inline lib/electric_slide/call_queue.rb]: I think this documentation is correct. The expected class name of the argument should be inside the square brackets, and the name of the variable should be outside, eg:

```
# @param [Hash] extra_params ...
```

Comment (rafbgarcia) [inline lib/electric_slide/call_queue.rb]: Ops! :P

I think I got confused on the above line `[Agent] agent`, but now it makes more sense :)

Thanks!

**Comment by bklang** [comment]
:+1:

**Pull request merged**